File: src/task/Scheduler/java/src/main/java/wniemiec/util/task/Scheduler.java
package wniemiec.util.task;

import java.util.*;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.ScheduledExecutorService;

/**
 * Responsible for calling routines after a certain time or whenever the the 
 * timer expires.
 * 
 * @author      William Niemiec &lt; williamniemiec@hotmail.com &gt;
 * 
 * @see         https://github.com/williamniemiec/utils/tree/master/src/task/Scheduler
 */
public class Scheduler {
	
	//-------------------------------------------------------------------------
	//		Attributes
	//-------------------------------------------------------------------------
	/**
	 * Stores routines sent to setInterval method along with its ids.
	 */
	private static final Map<Integer, Timer> intervalRoutines;
	
	/**
	 * Stores routines sent to setTimeout method along with its ids.
	 */
	private static final Map<Integer, Timer> delayRoutines;

	private static final Map<Long, Boolean> timeoutRoutines;
	private static final Map<Long, Future> timeoutRoutineThreads;
	private static long currentTimeoutRoutineId;


	//-------------------------------------------------------------------------
	//		Initialization blocks
	//-------------------------------------------------------------------------
	static {
		intervalRoutines = new HashMap<>();
		delayRoutines = new HashMap<>();
		timeoutRoutines = new HashMap<>();
		timeoutRoutineThreads = new HashMap<>();
	}

	
	//-------------------------------------------------------------------------
	//		Constructor
	//-------------------------------------------------------------------------
	private Scheduler() {
	}
	
	
	//-------------------------------------------------------------------------
	//		Methods
	//-------------------------------------------------------------------------
	/**
	 * Sets a timer which executes a routine once the timer expires.
	 * 
	 * @param		routine Routine to be performed
	 * @param		id Routine identifier (necessary to be able to stop it)
	 * @param		delay Waiting time before the routine is executed (in 
	 * milliseconds)
	 * 
	 * @return		False if an timeout has has already been set with the 
	 * given id or true otherwise
	 * 
	 * @throws		IllegalArgumentException If routine is null or if id or 
	 * delay is negative
	 */
	public static boolean setTimeout(Runnable routine, int id, long delay) {
		if (routine == null)
			throw new IllegalArgumentException("Routine cannot be null");
		
		if (id < 0)
			throw new IllegalArgumentException("Id cannot be negative");
		
		if (delay < 0)
			throw new IllegalArgumentException("Delay cannot be negative");
		
		if (delayRoutines.containsKey(id))
			return false;
		
		delayRoutines.put(id, scheduleTimeout(routine, delay));
		
		return true;
	}
	
	private static Timer scheduleTimeout(Runnable routine, long delay) {
		Timer timer = new Timer();
		timer.schedule(createTaskFromRoutine(routine), delay);
		
		return timer; 
	}

	private static TimerTask createTaskFromRoutine(Runnable routine) {
		return new TimerTask() {
		    @Override
		    public void run() {
		       routine.run();
		    }
		};
	}
	
	/**
	 * Repeatedly calls a routine with a fixed time delay between each call.
	 * 
	 * @param		routine Routine to be performed
	 * @param		id Routine identifier (necessary to be able to stop it)
	 * @param		interval Interval that the routine will be invoked (in 
	 * milliseconds)
	 * 
	 * @return		False if an interval has has already been set with the 
	 * given id or true otherwise
	 * 
	 * @throws		IllegalArgumentException If routine is null or if id or 
	 * interval is negative
	 */
	public static boolean setInterval(Runnable routine, int id, long interval) {
		if (routine == null)
			throw new IllegalArgumentException("Routine cannot be null");
		
		if (id < 0)
			throw new IllegalArgumentException("Id cannot be negative");
		
		if (interval < 0)
			throw new IllegalArgumentException("Interval cannot be negative");
		
		if (intervalRoutines.containsKey(id))
			return false;
		
		intervalRoutines.put(id, scheduleInterval(routine, interval));
		
		return true;
	}
	
	private static Timer scheduleInterval(Runnable routine, long interval) {
		Timer timer = new Timer();
		timer.scheduleAtFixedRate(createTaskFromRoutine(routine), 0, interval);
		
		return timer; 
	}
	
	/**
	 * Cancels a timed, repeating action, which was previously established by a
	 * call to {@link #setInterval(Runnable, int, long)}.
	 * 
	 * @param		id Routine id
	 */
	public static void clearInterval(int id) {
		if (!intervalRoutines.containsKey(id))
			return;
		
		intervalRoutines.get(id).cancel();
		intervalRoutines.remove(id);
	}
	
	/**
	 * Cancels a timed action which was previously established by a
	 * call to {@link #setTimeout(Runnable, int, long)}.
	 * 
	 * @param		id Routine id
	 */
	public static void clearTimeout(int id) {
		if (!delayRoutines.containsKey(id))
			return;
		
		delayRoutines.get(id).cancel();
		delayRoutines.remove(id);
	}
	
	public static void clearAllTimeout() {
		for (int timeoutId : delayRoutines.keySet()) {
			clearTimeout(timeoutId);
		}
	}
	
	public static void clearAllInterval() {
		for (int intervalId : intervalRoutines.keySet()) {
			clearInterval(intervalId);
		}
	}

	/**
	 * Runs a routine within a timeout. If the routine does not end on time, an
	 * interrupt signal will be sent to it.
	 *
	 * @param		routine Routine
	 * @param		timeout Maximum execution time (in milliseconds)
	 *
	 * @return		True if the routine has not finished executing within the
	 * time limit; false otherwise
	 *
	 * @throws IllegalArgumentException If routine is null or if timeout is
	 * is negative
	 */
	public static boolean setTimeoutToRoutine(Runnable routine, long timeout) {
		if (routine == null)
			throw new IllegalArgumentException("Routine cannot be null");

		if (timeout < 0)
			throw new IllegalArgumentException("Timeout cannot be negative");

		runRoutine(routine);
		waitRoutineFor(timeout);
		finishRoutine();

		return !hasRoutineFinished();
	}

	private static long getCurrentTime() {
		return new Date().getTime();
	}

	private static void runRoutine(Runnable routine) {
		initializeRoutineId();

		ScheduledExecutorService executor = Executors.newScheduledThreadPool(1);
		Future handler = executor.submit(() -> {
			routine.run();

			if (timeoutRoutines.containsKey(currentTimeoutRoutineId)) {
				timeoutRoutines.put(currentTimeoutRoutineId, true);
				timeoutRoutineThreads.remove(currentTimeoutRoutineId);
			}
		});
		executor.shutdown();

		timeoutRoutineThreads.put(currentTimeoutRoutineId, handler);
	}

	private static void initializeRoutineId() {
		currentTimeoutRoutineId = getCurrentTime();
		timeoutRoutines.put(currentTimeoutRoutineId, false);
	}

	private static void waitRoutineFor(long time) {
		if (currentTimeoutRoutineId < 0)
			return;

		long start = getCurrentTime();

		while ((timeElapsedInMilliseconds(start) < time) && !hasRoutineFinished()) {
			try {
				Thread.sleep(200);
			}
			catch (InterruptedException e) {
				break;
			}
		}
	}

	private static long timeElapsedInMilliseconds(long start) {
		return (getCurrentTime() - start);
	}

	private static boolean hasRoutineFinished() {
		if (currentTimeoutRoutineId < 0)
			return true;

		if (!timeoutRoutines.containsKey(currentTimeoutRoutineId))
			return false;

		return timeoutRoutines.get(currentTimeoutRoutineId);
	}

	private static void finishRoutine() {
		if (hasRoutineFinished())
			return;

		timeoutRoutines.remove(currentTimeoutRoutineId);
		timeoutRoutineThreads.get(currentTimeoutRoutineId).cancel(true);
		timeoutRoutineThreads.remove(currentTimeoutRoutineId);
	}
}

File: src/task/Scheduler/cpp/src/wniemiec/util/task/Scheduler.cpp
#include "../../../../include/wniemiec/util/task/Scheduler.hpp"

#include <ctime>
#include <chrono>
#include <thread>

using namespace wniemiec::util::task;

std::function<void(void)> null_routine = [](){};

//-------------------------------------------------------------------------
//		Attributes
//-------------------------------------------------------------------------
std::map<unsigned long, bool> Scheduler::delayRoutines = std::map<unsigned long, bool>();
std::map<unsigned long, bool> Scheduler::intervalRoutines = std::map<unsigned long, bool>();
std::map<unsigned long, bool> Scheduler::timeoutRoutine = std::map<unsigned lonf, bool>();
std::function<void(void)>& Scheduler::currentRoutine = null_routine;
unsigned long Scheduler::currentRoutineId;
pthread_t Scheduler::controlThread;


//-------------------------------------------------------------------------
//		Methods
//-------------------------------------------------------------------------
unsigned long Scheduler::set_timeout(const std::function<void(void)>& routine, long delay)
{
    initialize_routine_id();
    currentRoutine = routine;

    pthread_t thread;
    pthread_create(&thread, nullptr, delay_control_routine, (void*) delay);

    return currentRoutineId;
}

void Scheduler::initialize_routine_id()
{
    std::this_thread::sleep_for(std::chrono::seconds(1));
    currentRoutineId = get_current_time();
}

void* Scheduler::delay_control_routine(void* arg)
{
    long delay = (long) arg;
    unsigned long id = currentRoutineId;
    const std::function<void(void)> routine = currentRoutine;
    
    if (delay < 0)
        throw std::invalid_argument("Delay cannot be negative");

    delayRoutines[id] = true;

    std::this_thread::sleep_for(std::chrono::milliseconds(delay));

    if (delayRoutines[id])
        routine();
    
    return nullptr;
}

void Scheduler::clear_timeout(unsignedlong id)
{
    delayRoutines[id] = false;
}

unsigned long Scheduler::set_interval(const std::function<void(void)>& routine, long interval)
{
    initialize_routine_id();
    currentRoutine = routine;

    pthread_t thread;
    pthread_create(&thread, nullptr, interval_control_routine, (void*) interval);

    return currentRoutineId;
}

void* Scheduler::interval_control_routine(void* arg)
{
    long interval = (long) arg;
    unsigned long id = currentRoutineId;
    const std::function<void(void)> routine = currentRoutine;

    if (interval < 0)
        throw std::invalid_argument("Interval cannot be negative");
    
    intervalRoutines[id] = true;

    while (intervalRoutines[id])
    {
        std::this_thread::sleep_for(std::chrono::milliseconds(interval));
        routine();
    }
    
    return nullptr;
}

void Scheduler::clear_interval(unsigned long id)
{
    intervalRoutines[id] = false;
}

bool Scheduler::set_timeout_to_routine(const std::function<void(void)>& routine, long timeout)
{
    run_routine(routine);
    wait_routine_for(timeout);
    finish_routine();

    return !has_routine_finished();
}

void Scheduler::run_routine(const std::function<void(void)>& routine)
{
    initialize_routine_id();

    timeoutRoutine[currentRoutineId] = false;
    currentRoutine = routine;

    pthread_create(&controlThread, nullptr, control_routine, nullptr);
}

time_t Scheduler::get_current_time()
{
    return std::time(nullptr);
}

void* Scheduler::control_routine(void* args)
{
    time_t id = currentRoutineId;
    std::function<void(void)>& routine = currentRoutine;

    routine();
    timeoutRoutine[id] = true;
    
    return nullptr;
}

void Scheduler::wait_routine_for(long time)
{
    time_t start = get_current_time();

    while ((time_elapsed_in_milliseconds(start) < (double) time) && !has_routine_finished())
    {
        std::this_thread::sleep_for(std::chrono::milliseconds(200));
    }
}

double Scheduler::time_elapsed_in_milliseconds(time_t start)
{
    return (difftime(get_current_time(), start) * 1000);
}

bool Scheduler::has_routine_finished()
{
    return timeoutRoutine[currentRoutineId];
}

void Scheduler::finish_routine()
{
    pthread_cancel(controlThread);
}
